### PR TITLE
[wip] Update actix web to v4.0.0-beta.15

### PR DESCRIPTION
**Pull request by Jikstra**
Currently wip/draft as there's some things left to do:
- [x] fix `map_error_page()`
- [x] put http auth middleware back in again
- [x] somehow no files/folders are listed for me, i get `Route / could not be found` error
- [x] rebase
- [ ] wait for fix of https://github.com/actix/actix-web/pull/2398

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -149,18 +149,33 @@ impl MiniserveConfig {
...
             let mut keys =
                 pkcs8_private_keys(key_file).map_err(|_| anyhow!("Couldn't load private key"))?;
-            server_config.set_single_cert(cert_chain, keys.remove(0))?;
+            let server_config = rustls::ServerConfig::builder()
+                .with_safe_default_cipher_suites()
+                .with_safe_default_kx_groups()
+                .with_safe_default_protocol_versions()
+                .unwrap()
+                .with_no_client_auth()
+                .with_single_cert(vec!(rustls::Certificate(cert_chain)), rustls::PrivateKey(keys.remove(0)))?;
```
Comment (Jikstra) [inline src/config.rs:178]: I think I did well with refactoring this code, but please carefully review as I have not too much clue about tls things.

File: src/main.rs [hunk 2]
```diff
@@ -344,14 +345,14 @@ async fn favicon() -> impl Responder {
     let logo = include_str!("../data/logo.svg");
     HttpResponse::Ok()
         .insert_header(ContentType(mime::IMAGE_SVG))
-        .message_body(logo.into())
+        .message_body(BoxBody::new(logo))
```
Comment (aliemjay) [inline src/main.rs:348]:
```suggestion
        .body(logo)
```

File: src/listing.rs [hunk 3]
```diff
@@ -225,7 +225,7 @@ pub fn directory_listing(
                 .body(qr_to_svg_string(&qr, 2)),
             Err(err) => {
                 log::error!("URL is invalid (too long?): {:?}", err);
-                HttpResponse::UriTooLong().body(Body::Empty)
+                HttpResponse::UriTooLong().body(BoxBody::new(""))
```
Comment (aliemjay) [inline src/listing.rs:228]:
```suggestion
                HttpResponse::UriTooLong().finish() 
```

File: src/errors.rs [hunk 4]
```diff
@@ -141,14 +134,8 @@ where
...
-fn map_error_page(req: &HttpRequest, head: &mut ResponseHead, body: AnyBody) -> AnyBody {
-    let error_msg = match &body {
-        AnyBody::Bytes(bytes) => match std::str::from_utf8(bytes) {
-            Ok(msg) => msg,
-            _ => return body,
-        },
-        _ => return body,
-    };
+fn map_error_page(req: &HttpRequest, head: &mut ResponseHead, body: BoxBody) -> BoxBody {
+    let error_msg = "heeelp!";
```
Comment (aliemjay) [inline src/errors.rs:138]: I think we need to make this fn async and use actix_web::body::to_bytes

**Review by aliemjay — CHANGES_REQUESTED** [review]
Thanks for taking this forward.

File: src/errors.rs (hunk 4 above)
Comment (aliemjay) [reply src/errors.rs:138]: or alternatively use MessageBody::try_into_bytes

**Comment by Jikstra** [comment]
> Thanks for taking this forward.

Thanks for your initial pr :)

**Comment by Jikstra** [comment]
@aliemjay I think in your pr one test failed because of a regression in actix-web. To me it looked like the test was a bit anyways? Lot's of special chars if I remember correctly. Probably we can just kick out that test?

**Comment by aliemjay** [comment]
> Probably we can just kick out that test?

I don't think so. It is there for a reason and, if I remember, it fixed a reported issue. Anyway, it is on the plan for v4.

**Comment by aliemjay** [comment]
Oh, I forgot to note that most of my PR changes are now merged because I split it into a different PR. You may want to resolve the conflicts, but I think it is cleaner to rebase the last commits.

**Comment by Jikstra** [comment]
> Oh, I forgot to note that most of my PR changes are now merged because I split it into a different PR. You may want to resolve the conflicts, but I think it is cleaner to rebase the last commits.

I think I successfully rebased this pr :)

**Comment by Jikstra** [comment]
Soooo, all looks good now besides this one failing test. @aliemjay could we somehow encode those chars on our own in miniserve or do we need to wait for actix_web to fix this?

**Comment by aliemjay** [comment]
> do we need to wait for actix_web to fix this?

I believe yes. There is nothing we can do in our part. I'll push for it and try to get it soon.

**Comment by aliemjay** [comment]
Actix-web beta.19 is now out with a fix to the regression.

**Comment by Jikstra** [comment]
Awesome!

**Comment by svenstaro** [comment]
Hey, how are we looking on the fix? The actix-web 4 rc is out so now might be a good time to get this merged.

**Comment by Jikstra** [comment]
I will not have time to work on this before mid of march, so feel free to just continue this pr whoever wants to :)

**Pull request merged**